lightpath/controller.py

```diff
@@ -42,115 +50,403 @@ def __init__(self, client, endstations=None):
+                                                  target=device.md.name)
+                paths.extend(found_paths)
+            else:
+                logger.debug(f'No path between {src} and {device.md.name}')
 
-        except KeyError:
-            raise ValueError("Beamline {} not found".format(bl.name))
+        if not paths:
+            raise PathError(f'No paths from sources to {device.md.name}')
+
+        subgraphs = [self.graph.subgraph(p) for p in paths]
+        beampaths = []
+        for subg in subgraphs:
+            devs = [node[1]['dev'] for node in subg.nodes.data()
+                    if node[1]['dev'] is not None]
+            beampaths.append(BeamPath(*devs, name=f'{device.md.name}_path'))
+
+        paths_by_length = sorted(beampaths, key=self.imped_z)
+
+        return paths_by_length[-1]
+
+    @staticmethod
+    def make_graph(
+        branch_devs: List[SearchResult],
+        branch_name: str,
+        sources: List[NodeName] = []
+    ) -> nx.DiGraph:
+        """
+        Create a graph with devices from branch_devs as nodes,
+        arranged in z-order.
+
+        It is assumed that all devices lie on branch: branch_name
+
+        If sources is provided, will prepend a source node at the
+        beginning of the branch when branch_name == sources
+
+        Parameters
+        ----------
+        branch_devs : List[happi.SearchResult]
+            a list of devices to generate graph with
 
-        return prior
+        branch_name : str
+            branch name, used to label edges
+
+        sources : List[NodeName], optional
+            source node Names, by default []
+
+        Returns
+        -------
+        nx.DiGraph
+            The branch comprised of nodes holding devices from branch_devs
+        """
+        graph = nx.DiGraph()
+        result_list = list(branch_devs)
+        result_list.sort(key=lambda x: x.metadata['z'])
+        # label nodes with device name, store ophyd device
+        nodes = []
+        for res in result_list:
+            try:
+                dev = res.get()
+            except Exception:
+                # TODO: be better about specific exceptions
+                logger.debug(
+                    f'Failed to initialize device: {res["name"]}'
+                )
+                continue
+            nodes.append((res.metadata['name'], {'dev': dev}))
+
+        # construct edges
+        edges: List[Tuple[NodeName, NodeName, Dict[str, Any]]] = []
+        skipped_right: List[int] = []
+        skipped_left: List[int] = []
+        last_on_branch = 0
+        # weight not used currently, but may be for path finding algos
+        edata = {'weight': 0.0, 'branch': branch_name}
+        # Need to properly handle devices that either:
+        # only have the current branch in their input (skipped_right)
+        # only have the current branch in their output (skipped_left)
+        for i in range(len(nodes)):
```

Comment:
This one was a tougher one for me to figure out as well, I'll try to clarify in the comments.  

This logic is complicated because it's trying to handle "dangling" nodes.  These nodes either only have the current branch as an input or only as an output.  These nodes are distinct from branching devices, and I had to make sure only the correct edges were constructed.  

I couldn't figure out a concise way to describe the logic, but maybe something more concise would be more confusing to read through